### PR TITLE
[pytorch] Add strong Wolfe line search for lbfgs

### PR DESCRIPTION
**Pull request by fehiepsi**
This pull request adds a line search for lbfgs. "strong Wolfe" is the default line search method in [minFunc](https://www.cs.ubc.ca/~schmidtm/Software/minFunc.html) and it is also recommended in the [Numerical Optimization](https://www.springer.com/gp/book/9780387303031) book.

The implementation is based on four sources:
+ https://www.cs.ubc.ca/~schmidtm/Software/minFunc.html
+ https://www.springer.com/gp/book/9780387303031 Algorithms 3.5, 3.6, formula 3.59
+ https://github.com/torch/optim/blob/master/lswolfe.lua
+ https://github.com/torch/optim/blob/master/polyinterp.lua

The 'lua' version is based on an old version of `minFunc`, which has been updated in 2012. I made a couple of small changes based on the updated version. Due to that, the test of comparing with `.lua` version is not consistent (that's is the reason I changed a learning rate in the test).

Differential Revision: [D15740107](https://our.internmc.facebook.com/intern/diff/15740107/)

### CODE REVIEW

File: torch/optim/lbfgs.py [hunk 1]
```diff
@@ -146,10 +321,12 @@ def step(self, closure):
                         # shift history by one (limited-memory)
                         old_dirs.pop(0)
                         old_stps.pop(0)
+                        ro.pop(0)
 
                     # store new direction/step
                     old_dirs.append(y)
                     old_stps.append(s)
+                    ro.append(1. / ys)
```
Comment (fehiepsi) [inline torch/optim/lbfgs.py]: by this, we don't have to recalculate ro

File: torch/optim/lbfgs.py [hunk 2]
```diff
@@ -191,18 +363,31 @@ def step(self, closure):
...
+                t = min(1., 1. / flat_grad.abs().sum()) * lr
             else:
                 t = lr
 
             # directional derivative
             gtd = flat_grad.dot(d)  # g * d
 
+            # directional derivative is below tolerance
+            if gtd > -tolerance_change:
+                break
```
Comment (fehiepsi) [inline torch/optim/lbfgs.py]: the position of this check is from minFunc 2012 version

File: torch/optim/lbfgs.py [hunk 3]
```diff
@@ -228,13 +413,12 @@ def step(self, closure):
...
-                break
-
-            if gtd > -tolerance_change:
+            # optimal condition
+            if opt_cond:
                 break
 
-            if d.mul(t).abs_().sum() <= tolerance_change:
+            # lack of progress
+            if d.mul(t).abs().max() <= tolerance_change:
```
Comment (fehiepsi) [inline torch/optim/lbfgs.py]: the reason for change is similar to opt_cond

File: torch/optim/lbfgs.py [hunk 4]
```diff
@@ -106,16 +278,18 @@ def step(self, closure):
         state['func_evals'] += 1
 
         flat_grad = self._gather_flat_grad()
-        abs_grad_sum = flat_grad.abs().sum()
+        opt_cond = flat_grad.abs().max() <= tolerance_grad
```
Comment (fehiepsi) [inline torch/optim/lbfgs.py]: optimal condition is based on abs_grad_max instead of abs_grad_sum, this comes from minFunc 2012 (note that minFunc 2009 used abs_grad_sum)

File: torch/optim/lbfgs.py [hunk 5]
```diff
@@ -3,8 +3,165 @@
...
+    d2_square = d1 ** 2 - g1 * g2
+    if d2_square >= 0:
+        d2 = d2_square.sqrt()
+        if x1 <= x2:
+            min_pos = x2 - (x2 - x1) * ((g2 + d2 - d1) / (g2 - g1 + 2 * d2))
+        else:
+            min_pos = x1 - (x1 - x2) * ((g1 + d2 - d1) / (g1 - g2 + 2 * d2))
+        return min(max(min_pos, xmin_bound), xmax_bound)
+    else:
+        return (xmin_bound + xmax_bound) / 2.
```
Comment (fehiepsi) [inline torch/optim/lbfgs.py]: I revised the implementation a bit (comparing to both lua and minFunc's implementation).

File: test/test_optim.py [hunk 6]
```diff
@@ -467,13 +467,17 @@ def test_lbfgs(self):
             wrap_old_fn(old_optim.lbfgs)
         )
         self._test_rosenbrock(
-            lambda params: optim.LBFGS(params, lr=5e-2, max_iter=5),
-            wrap_old_fn(old_optim.lbfgs, learningRate=5e-2, maxIter=5)
+            lambda params: optim.LBFGS(params, lr=1, max_iter=5),
+            wrap_old_fn(old_optim.lbfgs, learningRate=1, maxIter=5)
```
Comment (fehiepsi) [inline test/test_optim.py]: conditions change, so the test is not the same as before. Anyway, with lr=1, the test passed.

**Comment by SsnL** [comment]
I'll review this soon.

File: test/test_optim.py (hunk 6 above)
Comment (sri123098) [reply test/test_optim.py]: Could you please elaborate on "why are you increasing the learning rate"?

Comment (fehiepsi) [reply test/test_optim.py]: This test compares to the `lua` version. I make some small changes based on the updated version of `minFunc`, so the results of `pytorch` and `lua` versions are a little bit different.

**Comment by ailzhang** [comment]
cc: @SsnL

File: torch/optim/lbfgs.py [hunk 7]
```diff
@@ -3,8 +3,165 @@
...
+    #   t_new = min(max(min_pos,xmin_bound),xmax_bound);
+    d1 = g1 + g2 - 3 * (f1 - f2) / (x1 - x2)
+    d2_square = d1 ** 2 - g1 * g2
+    if d2_square >= 0:
+        d2 = d2_square.sqrt()
+        if x1 <= x2:
+            min_pos = x2 - (x2 - x1) * ((g2 + d2 - d1) / (g2 - g1 + 2 * d2))
+        else:
+            min_pos = x1 - (x1 - x2) * ((g1 + d2 - d1) / (g1 - g2 + 2 * d2))
+        return min(max(min_pos, xmin_bound), xmax_bound)
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: As mentioned before equation 3.59, the minimum of the cubic function will occur either at min_pos, xmin_bound, or xmax_bound. Since equation 3.59 doesn’t provide the value of the cubic polynomial at the minimum, and computing it might increase the computational cost, we could simply return xmin_bound or xmax_bound depending on which one has a lower value of f, whenever min_pos falls outside the range of xmin_bound and xmax_bound.

An example (which could be used for testing this algorithm) would be interpolating the function f(x)=-x(x-2)**2 on [0,1]. The interpolation is exact, since it is a cubic polynomial. The values are f(0)=0, f(1)=1, g(0)=f'(0)=4, g(1)=f'(1)=5. The local minimum would be found at x=2 outside the range [0,1]. This algorithm would return x=1 since that is the closest point to the range, but the minimum on this range is at 0.

That being said, this is indeed implementing what is done in https://github.com/torch/optim/blob/master/polyinterp.lua#L131, so I am okay with using either as authoritative to move ahead, and leaving this as a later improvement.

File: torch/optim/lbfgs.py [hunk 8]
```diff
@@ -3,8 +3,165 @@
...
+            bracket_g = [g_new]
+            done = True
+            break
+
+        if gtd_new >= 0:
+            bracket = [t_prev, t]
+            bracket_f = [f_prev, f_new]
+            bracket_g = [g_prev, g_new.clone()]
+            bracket_gtd = [gtd_prev, gtd_new]
+            break
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: Can you give a reference for the case gtd_new >= 0? Algorithm 35 switches the bounds of the bracket, and so I would have expected

            bracket = [t, t_prev]
            bracket_f = [f_new, f_prev]
            bracket_g = [g_new.clone(), g_prev]
            bracket_gtd = [gtd_new, gtd_prev]

File: torch/optim/lbfgs.py (hunk 5 above)
Comment (vincentqb) [inline torch/optim/lbfgs.py]: Same comment as above about which value to pick when local minimum is outside of the range of xmin_bound and xmax_bound.

File: torch/optim/lbfgs.py [hunk 9]
```diff
@@ -3,8 +3,165 @@
...
+    # exact point satisfying the criteria
+    insuf_progress = False
+    # find high and low points in bracket
+    low_pos, high_pos = (0, 1) if bracket_f[0] <= bracket_f[-1] else (1, 0)
+    while not done and ls_iter < max_ls:
+        # compute new trial value
+        t = _cubic_interpolate(bracket[0], bracket_f[0], bracket_gtd[0],
+                               bracket[1], bracket_f[1], bracket_gtd[1])
+
+        # test what we are making sufficient progress
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: I expect this block safeguards against having the new point too close to the boundary. Can you confirm this is indeed so? Then I would update this comment to make it more descriptive.

File: torch/optim/lbfgs.py (hunk 7 above)
Comment (fehiepsi) [reply torch/optim/lbfgs.py]: I agree that it might be better to compare f1, f2 and pick x_i with smaller f_i when min_pos is outside of the interval. But I would like to stick with original implementations, and leave this for later improvement.

> Does minFunc also do as in lua?

Yes, minFunc also does so.

File: torch/optim/lbfgs.py (hunk 8 above)
Comment (fehiepsi) [reply torch/optim/lbfgs.py]: If I understand correctly, then we don't require bracket has some sort of order because we will choose `low_pos`, `high_pos` before the zooming phase.

File: torch/optim/lbfgs.py (hunk 9 above)
Comment (fehiepsi) [reply torch/optim/lbfgs.py]: Yes, this is intended to do so. I'll add more comments for this behaviour.

File: torch/optim/lbfgs.py (hunk 8 above)
Comment (vincentqb) [reply torch/optim/lbfgs.py]: The sequence of {t_i} (or {alpha_i} in the book) is monotonically increasing, but the inputs to zoom may not be, see comment after Algorithm 3.5. In Algorithm 3.6, alpha_low and alpha_high do not refer to the smaller and bigger one.

Are you saying line 99 compensate for the case covered here, in lines 64-69? Line 99 is absent from Algorithm 3.6.

File: torch/optim/lbfgs.py [hunk 10]
```diff
@@ -3,8 +3,165 @@
...
+        bracket = [0, t]
+        bracket_f = [f, f_new]
+        bracket_g = [g, g_new]
+
+    # zoom phase: we now have a point satisfying the criteria, or
+    # a bracket around it. We refine the bracket until we find the
+    # exact point satisfying the criteria
+    insuf_progress = False
+    # find high and low points in bracket
+    low_pos, high_pos = (0, 1) if bracket_f[0] <= bracket_f[-1] else (1, 0)
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: See discussion for lines 64-69.

File: torch/optim/lbfgs.py (hunk 8 above)
Comment (fehiepsi) [reply torch/optim/lbfgs.py]: I think that bracket is only used for `zoom` and quoted from the book, alpha_low is
> among all step lengths generated so far and satisfying the sufficient decrease condition, the one giving the smallest function value

The line 99 guarantees that condition. To me, both orders are equivalent but I can revert the order as you commented to avoid confusion. What do you think?

Comment (vincentqb) [reply torch/optim/lbfgs.py]: Indeed :) Thanks for clarifying!

**Comment by vincentqb** [comment]
Besides one test failing, this looks good to me :)

**Comment by fehiepsi** [comment]
Thanks for reviewing @vincentqb !

**Comment by vincentqb** [comment]
Thanks for the references! This PR looks good to me :)

**Review by vincentqb — APPROVED** [review]
Thanks for the references! This PR looks good to me :)
